runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -448,8 +457,18 @@ function M._enable(bufnr)
 
     api.nvim_create_autocmd('LspDetach', {
       buffer = bufnr,
-      callback = function()
-        disable(bufnr)
+      callback = function(args)
+        local remaining_client_count = 0
+        for _, client in
+          pairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_diagnostic }))
+        do
+          if client.id ~= args.data.client_id then
+            remaining_client_count = remaining_client_count + 1
+          end
+        end
+        if remaining_client_count == 0 then
+          disable(bufnr)
+        end
```

Comment:
vim.iter may be worth using here (and in inlay_hint.lua).

```lua
local clients = vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_diagnostic })
local remaining = vim.iter(clients):fold(0, function(acc, c)
  return (c.id == args.data.client_id) and i or i + 1
end)

if remaining == 0 then
  disable(bufnr)
end
```

Reply:
> vim.iter may be worth using here (and in inlay_hint.lua).

Maybe like this?

```lua
local clients = vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_diagnostic })
if vim.iter(clients):any(function(client) return client.id ~= args.data.client_id end) then
  disable(bufnr)
end
```

Reply:
That's much nicer, thank you!